### PR TITLE
Fix flaky test_close_gracefully and test_lifetime

### LINKED ISSUE #4862: Flaky distributed/tests/test_worker.py::test_lifetime
Random test failure of `distributed/tests/test_worker.py::test_lifetime`

First observed in https://github.com/dask/distributed/pull/4774/checks?check_run_id=2693532608

<details/>

```
    @pytest.mark.slow
    @pytest.mark.asyncio
    async def test_lifetime(cleanup):
        async with Scheduler() as s:
            async with Worker(s.address) as a, Worker(s.address, lifetime="1 seconds") as b:
                async with Client(s.address, asynchronous=True) as c:
                    futures = c.map(slowinc, range(200), delay=0.1)
                    await asyncio.sleep(1.5)
                    assert b.status != Status.running
                    await b.finished()
    
>                   assert set(b.data).issubset(a.data)  # successfully moved data over
E                   assert False
E                    +  where False = <built-in method issubset of set object at 0x7f8134ca3ba0>(Buffer<<LRU: 1176/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-bi3offm0/storage, mode="a", 0 elements>>>)
E                    +    where <built-in method issubset of set object at 0x7f8134ca3ba0> = {'slowinc-0093af3b7e8e6f16b973e417caa4f85c', 'slowinc-120b294a681d6c5fc1bb125b7dc27a53', 'slowinc-28ec16f2ca5777e76e1e...eeed2f63f7030bb5542ededd', 'slowinc-4b9f7bf1e74f50ba387496336a3ab2de', 'slowinc-4eb81a10f2358a764e70dd9443ac65a8', ...}.issubset
E                    +      where {'slowinc-0093af3b7e8e6f16b973e417caa4f85c', 'slowinc-120b294a681d6c5fc1bb125b7dc27a53', 'slowinc-28ec16f2ca5777e76e1e...eeed2f63f7030bb5542ededd', 'slowinc-4b9f7bf1e74f50ba387496336a3ab2de', 'slowinc-4eb81a10f2358a764e70dd9443ac65a8', ...} = set(Buffer<<LRU: 588/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-5762snfd/storage, mode="a", 0 elements>>>)
E                    +        where Buffer<<LRU: 588/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-5762snfd/storage, mode="a", 0 elements>>> = <Worker: 'tcp://10.212.11.232:52856', tcp://10.212.11.232:52856, Status.closed, stored: 21, running: 3/3, ready: 76, comm: 0, waiting: 0>.data
E                    +    and   Buffer<<LRU: 1176/9019431321 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /Users/runner/work/distributed/distributed/dask-worker-space/worker-bi3offm0/storage, mode="a", 0 elements>>> = <Worker: 'tcp://10.212.11.232:52850', tcp://10.212.11.232:52850, Status.running, stored: 42, running: 3/3, ready: 155, comm: 0, waiting: 0>.data

distributed/tests/test_worker.py:1588: AssertionError
```

### PR DESCRIPTION
**Pull request by crusaderky**
Closes #4862

### CODE REVIEW

File: distributed/tests/test_worker.py [hunk 1]
```diff
@@ -1627,32 +1627,52 @@ async def test_worker_listens_on_same_interface_by_default(cleanup, Worker):
...
 
 
 @pytest.mark.slow
 @gen_cluster(client=True, nthreads=[])
 async def test_lifetime(c, s):
     async with Worker(s.address) as a, Worker(s.address, lifetime="1 seconds") as b:
-        futures = c.map(slowinc, range(200), delay=0.1, worker=[b.address])
-        await asyncio.sleep(1.5)
-        assert b.status not in (Status.running, Status.paused)
+        futures = c.map(slowinc, range(200), delay=0.1, workers=[b.address])
```
Comment (crusaderky) [inline distributed/tests/test_worker.py:1667]: fix spelling error worker->workers!

**Comment by crusaderky** [comment]
All test failures are unrelated

File: distributed/tests/test_worker.py [hunk 2]
```diff
@@ -1627,32 +1627,52 @@ async def test_worker_listens_on_same_interface_by_default(cleanup, Worker):
...
 @gen_cluster(client=True, nthreads=[])
 async def test_lifetime(c, s):
-    async with Worker(s.address) as a, Worker(s.address, lifetime="1 seconds") as b:
-        futures = c.map(slowinc, range(200), delay=0.1, worker=[b.address])
-        await asyncio.sleep(1.5)
-        assert b.status not in (Status.running, Status.paused)
+    async with Worker(s.address) as a, Worker(s.address, lifetime="2 seconds") as b:
+        futures = c.map(slowinc, range(200), delay=0.1, workers=[b.address])
+        await asyncio.sleep(1)
+        assert not a.data
```
Comment (crusaderky) [inline distributed/tests/test_worker.py:1669]: After fixing the misspelling above, the test was occasionally failing this assertion with ``lifetime="1 seconds"`` and ``sleep(0.5)``

**Comment by crusaderky** [comment]
All test failures are unrelated; ready for review and merge

File: distributed/tests/test_worker.py [hunk 3]
```diff
@@ -1627,32 +1627,68 @@ async def test_worker_listens_on_same_interface_by_default(cleanup, Worker):
...
+    while True:
+        mem = {k for k, ts in s.tasks.items() if ts.state == "memory"}
+        if len(mem) >= 8:
+            break
         await asyncio.sleep(0.01)
-    mem = set(b.data)
-    proc = {ts for ts in b.tasks.values() if ts.state == "executing"}
-    assert proc
+
+    assert any(ts for ts in b.tasks.values() if ts.state == "executing")
```
Comment (fjetter) [inline distributed/tests/test_worker.py:1642]: I think there is a tiny time window possible where no task is in `executing` state until the next `ensure_computing` is called

File: distributed/tests/test_worker.py [hunk 4]
```diff
@@ -1627,32 +1627,68 @@ async def test_worker_listens_on_same_interface_by_default(cleanup, Worker):
...
-        assert ts.state in ("executing", "memory")
+
+    # All tasks that were in memory in b have been copied over to a;
+    # they have not been recomputed
+    for key in mem:
+        assert_worker_story(
+            a.story(key),
+            [
+                (key, "put-in-memory"),
+                (key, "receive-from-scatter"),
```
Comment (fjetter) [inline distributed/tests/test_worker.py:1656]: why is this a scatter event?

File: distributed/tests/test_worker.py (hunk 1 above)
Comment (fjetter) [reply distributed/tests/test_worker.py:1667]: something like `await next(as_completed(futures))` would protect us there and may remove the necessity for sleep below. I'm fine without it

File: distributed/tests/test_worker.py (hunk 4 above)
Comment (crusaderky) [reply distributed/tests/test_worker.py:1656]: legacy (pre-AMM) Scheduler.gather_on_worker, which underlies both Scheduler.replicate and Scheduler.retire_workers

File: distributed/tests/test_worker.py (hunk 1 above)
Comment (crusaderky) [reply distributed/tests/test_worker.py:1667]: Revised

**Comment by crusaderky** [comment]
All review comments have been incorporated.
All test failures are unrelated.

**Review by fjetter — APPROVED** [review]

**Pull request merged**